### ISSUE TITLE
Curl class not calls __destruct when out of scope

### ISSUE THREAD

**Issue by IvixLabs** [php-curl-class/php-curl-class#307]
Seem problem in closure function inside functions setDefaultJsonDecoder and setDefaultXmlDecoder.

I do not know maybe it is php bug.

```php
class MyCurl {

    public $curl;

    private $closure;

    /**
     * MyCurl constructor.
     */
    public function __construct()
    {
        $this->curl = curl_init();
        $this->closure = function() {

        };
    }

    function __destruct()
    {
        var_dump('Destructor calling');
    }
}
```

this class not call destructor

```php
class MyCurl {

    public $curl;

    /**
     * MyCurl constructor.
     */
    public function __construct()
    {
        $this->curl = curl_init();
    }

    function __destruct()
    {
        var_dump('Destructor calling');
    }
}
```

this call.

Otherwise i have too much opened connections in php worker(not shutdown each time):

` error:02001018:system library:f0pen:Too many open files [] []`

I fixed problem by set setOpt(CURLOPT_FORBID_REUSE, true);

But i think better remove closure functions maybe make some object with interface.

Thanks.

**Comment by zachborboa** [comment]
To reproduce the `Too many open files` error:

```php
use \Curl\MultiCurl;

$m = new MultiCurl();
for ($i = 0; $i <= 300; $i++) {
    $m->addDownload('https://www.example.com/', 'file' . $i);
}
```

> PHP Warning:  fopen(filename): failed to open stream: Too many open files in ...

**Comment by zachborboa** [comment]
Hi. The too many open files was fixed in #375. I'm also seeing `__destruct()` being called as expected.

```bash
$ php test.php 
__construct1
__construct2
__destruct2
__destruct1
````

```php
// test.php
<?php
class MyCurl1 {
    public $curl;
    private $closure;

    public function __construct()
    {
        echo '__construct1' . "\n";
        $this->curl = curl_init();
        $this->closure = function() {
        };
    }

    function __destruct()
    {
        echo '__destruct1' . "\n";
    }
}

class MyCurl2 {
    public $curl;

    public function __construct()
    {
        echo '__construct2' . "\n";
        $this->curl = curl_init();
    }

    function __destruct()
    {
        echo '__destruct2' . "\n";
    }
}

$a = new MyCurl1();
$b = new MyCurl2();
```
